plugins/trace_api_plugin/abi_data_handler.cpp

```diff
@@ -29,6 +29,29 @@ namespace eosio::trace_api {
          }
       }
 
+      return {};
+   }
+
+      fc::variant abi_data_handler::process_data(const action_trace_v1& action, const yield_function& yield ) {
+      if (abi_serializer_by_account.count(action.account) > 0) {
+         const auto& serializer_p = abi_serializer_by_account.at(action.account);
+         auto type_name = serializer_p->get_action_type(action.action);
+
+         if (!type_name.empty()) {
+            try {
+               // abi_serializer expects a yield function that takes a recursion depth
+               auto abi_yield = [yield](size_t recursion_depth) {
+                  yield();
+                  EOS_ASSERT( recursion_depth < chain::abi_serializer::max_recursion_depth, chain::abi_recursion_depth_exception,
+                              "exceeded max_recursion_depth ${r} ", ("r", chain::abi_serializer::max_recursion_depth) );
+               };
+               return serializer_p->binary_to_variant(type_name, action.data, abi_yield);
+            } catch (...) {
+               except_handler(MAKE_EXCEPTION_WITH_CONTEXT(std::current_exception()));
+            }
+         }
+      }
```

Comment:
Can we make this a template or have it take a `std::variant`  so all this code is not duplicated.

Reply:
yes, That is what mean to decouple structures. I'm going to use std::variant and will give a new pull later. Thanks.

Reply:
The new pull was https://github.com/EOSIO/eos/pull/9269, which was made weeks ago,  I just reform the structures so as to save work hours for future versions changes. As others didn't want more changes, so I didn't use pull 9629 just continue based on this pull 9261.

Reply:
Good idea to combine process_data() and process_return(), will do it later.

Reply:
Already combined the  process_data() and process_return().